app/org/sagebionetworks/bridge/services/ParticipantService.java

```diff
@@ -158,6 +160,87 @@ public StudyParticipant getParticipant(Study study, String email) {
         }
         return accountDao.getPagedAccountSummaries(study, offsetBy, pageSize, emailFilter);
     }
+
+    /**
+     * Create a study participant. A password must be provided, even if it is added on behalf of a 
+     * user before triggering a reset password request.  
+     */
+    public void createParticipant(Study study, StudyParticipant participant) {
+        checkNotNull(study);
+        checkNotNull(participant);
+        
+        Validate.entityThrowingException(new StudyParticipantValidator(study, true), participant);
+
+        externalIdService.reserveExternalId(study, participant.getExternalId());
+        
+        SignUp signUp = new SignUp(participant.getEmail(), participant.getPassword(), null, null);
+        Account account = accountDao.signUp(study, signUp, study.isEmailVerificationEnabled());
+        
+        // This should not fail as we're now creating the mapping in sign-up itself. User's email
+        // doesn't even need to be verified before we can add this information to the account.
+        HealthId healthId = healthCodeService.getMapping(account.getHealthId());
+        String healthCode = healthId.getCode();
+        
+        Map<ParticipantOption,String> options = Maps.newHashMap();
+        for (ParticipantOption option : ParticipantOption.values()) {
+            options.put(option, option.fromParticipant(participant));
+        }
+        if (study.isExternalIdValidationEnabled()) {
+            options.remove(EXTERNAL_IDENTIFIER);
+        }
+        optionsService.setAllOptions(study.getStudyIdentifier(), healthCode, options);
+        
+        account.setFirstName(participant.getFirstName());
+        account.setLastName(participant.getLastName());
+        for(String attribute : study.getUserProfileAttributes()) {
+            String value = participant.getAttributes().get(attribute);
+            account.setAttribute(attribute, value);
+        }
+        accountDao.updateAccount(study, account);
+        
+        externalIdService.assignExternalId(study, participant.getExternalId(), healthCode);
+    }
+    
+    public void updateParticipant(Study study, String email, StudyParticipant participant) {
```

Comment:
External IDs have some rules, if we are managing them (if not, it's just a string you can add/update however the client wants):

- you can add them at creation or on update, but once added, they cannot be changed or removed.
- an ID should be assigned to one and only one person. The external ID table is used to make sure we don't write the same ID to two different records in the ParticipantOptions table.